File: controller/scaler.py
"""
Autoscaling logic for HTTP applications.
Makes scaling decisions based on metrics like RPS, latency, CPU, and memory.
"""

import time
import logging
import statistics
from typing import Dict, List, Optional
from dataclasses import dataclass, field
from collections import deque, defaultdict

logger = logging.getLogger(__name__)

@dataclass
class ScalingPolicy:
    """Scaling policy configuration for an application."""
    min_replicas: int = 1
    max_replicas: int = 5
    target_rps_per_replica: int = 50
    max_p95_latency_ms: int = 250
    max_conn_per_replica: int = 80
    scale_out_threshold_pct: int = 80
    scale_in_threshold_pct: int = 30
    window_seconds: int = 20
    cooldown_seconds: int = 30
    max_cpu_percent: float = 70.0
    max_memory_percent: float = 75.0

@dataclass
class MetricPoint:
    """A single metric measurement."""
    timestamp: float
    value: float

@dataclass
class ScalingMetrics:
    """Container metrics used for scaling decisions."""
    rps: float = 0.0
    p95_latency_ms: float = 0.0
    active_connections: int = 0
    cpu_percent: float = 0.0
    memory_percent: float = 0.0
    healthy_replicas: int = 0
    total_replicas: int = 0

@dataclass
class ScalingDecision:
    """Result of a scaling evaluation."""
    should_scale: bool
    target_replicas: int
    current_replicas: int
    reason: str
    triggered_by: List[str] = field(default_factory=list)
    metrics: Optional[ScalingMetrics] = None

class AutoScaler:
    def __init__(self):
        self.policies: Dict[str, ScalingPolicy] = {}
        self.metrics_history: Dict[str, Dict[str, deque]] = defaultdict(lambda: defaultdict(lambda: deque(maxlen=1000)))
        self.last_scale_time: Dict[str, float] = {}
        self.scale_decisions: Dict[str, deque] = defaultdict(lambda: deque(maxlen=100))
        # Store last calculated scale factors for debug/inspection
        self.last_scale_factors: Dict[str, Dict[str, float]] = {}
        
    def set_policy(self, app_name: str, policy: ScalingPolicy):
        """Set the scaling policy for an application."""
        self.policies[app_name] = policy
        logger.info(f"Set scaling policy for {app_name}: min={policy.min_replicas}, max={policy.max_replicas}")
    
    def get_policy(self, app_name: str) -> Optional[ScalingPolicy]:
        """Get the scaling policy for an application."""
        return self.policies.get(app_name)
    
    def add_metrics(self, app_name: str, metrics: ScalingMetrics):
        """Add new metrics for an application."""
        timestamp = time.time()
        
        # Store metrics with timestamp
        history = self.metrics_history[app_name]
        history["rps"].append(MetricPoint(timestamp, metrics.rps))
        history["latency"].append(MetricPoint(timestamp, metrics.p95_latency_ms))
        history["connections"].append(MetricPoint(timestamp, metrics.active_connections))
        history["cpu"].append(MetricPoint(timestamp, metrics.cpu_percent))
        history["memory"].append(MetricPoint(timestamp, metrics.memory_percent))
        history["healthy_replicas"].append(MetricPoint(timestamp, metrics.healthy_replicas))
        
        # Clean old metrics
        self._clean_old_metrics(app_name, timestamp)
    
    def _clean_old_metrics(self, app_name: str, current_time: float):
        """Remove metrics older than the policy window."""
        policy = self.policies.get(app_name)
        if not policy:
            return
        
        cutoff_time = current_time - (policy.window_seconds * 3)  # Keep 3x window for analysis
        
        for metric_type, points in self.metrics_history[app_name].items():
            while points and points[0].timestamp < cutoff_time:
                points.popleft()
    
    def evaluate_scaling(self, app_name: str, current_replicas: int) -> ScalingDecision:
        """Evaluate if scaling is needed for an application."""
        policy = self.policies.get(app_name)
        if not policy:
            return ScalingDecision(
                should_scale=False,
                target_replicas=current_replicas,
                current_replicas=current_replicas,
                reason="No scaling policy configured"
            )
        
        # Check cooldown period
        last_scale = self.last_scale_time.get(app_name, 0)
        if time.time() - last_scale < policy.cooldown_seconds:
            return ScalingDecision(
                should_scale=False,
                target_replicas=current_replicas,
                current_replicas=current_replicas,
                reason=f"In cooldown period ({policy.cooldown_seconds}s)"
            )
        
        # Get recent metrics
        metrics = self._get_recent_metrics(app_name, policy.window_seconds)
        if not metrics:
            return ScalingDecision(
                should_scale=False,
                target_replicas=current_replicas,
                current_replicas=current_replicas,
                reason="No recent metrics available"
            )
        
        # Calculate scaling factors for each metric
        scale_factors = self._calculate_scale_factors(metrics, policy)
        # Save for external debugging/metrics endpoint
        self.last_scale_factors[app_name] = scale_factors
        logger.debug(f"[autoscale] {app_name} metrics={metrics} scale_factors={scale_factors}")
        
        # Determine if we should scale out or in
        decision = self._make_scaling_decision(
            app_name, current_replicas, scale_factors, policy, metrics
        )
        
        # Record the decision
        self.scale_decisions[app_name].append(decision)
        
        return decision
    
    def _get_recent_metrics(self, app_name: str, window_seconds: int) -> Optional[ScalingMetrics]:
        """Get aggregated metrics for the recent window."""
        current_time = time.time()
        cutoff_time = current_time - window_seconds
        
        history = self.metrics_history[app_name]
        
        # Get recent points for each metric
        recent_rps = [p.value for p in history["rps"] if p.timestamp >= cutoff_time]
        recent_latency = [p.value for p in history["latency"] if p.timestamp >= cutoff_time]
        recent_connections = [p.value for p in history["connections"] if p.timestamp >= cutoff_time]
        recent_cpu = [p.value for p in history["cpu"] if p.timestamp >= cutoff_time]
        recent_memory = [p.value for p in history["memory"] if p.timestamp >= cutoff_time]
        recent_healthy = [p.value for p in history["healthy_replicas"] if p.timestamp >= cutoff_time]
        
        if not recent_rps and not recent_latency:
            return None
        
        # Aggregate metrics (using mean for most, max for latency)
        return ScalingMetrics(
            rps=statistics.mean(recent_rps) if recent_rps else 0.0,
            p95_latency_ms=max(recent_latency) if recent_latency else 0.0,
            active_connections=int(statistics.mean(recent_connections)) if recent_connections else 0,
            cpu_percent=statistics.mean(recent_cpu) if recent_cpu else 0.0,
            memory_percent=statistics.mean(recent_memory) if recent_memory else 0.0,
            healthy_replicas=int(statistics.mean(recent_healthy)) if recent_healthy else 1,
            total_replicas=int(statistics.mean(recent_healthy)) if recent_healthy else 1
        )
    
    def _calculate_scale_factors(self, metrics: ScalingMetrics, policy: ScalingPolicy) -> Dict[str, float]:
        """Calculate scaling factors for each metric (1.0 = at target, >1.0 = overloaded)."""
        factors = {}
        
        if metrics.healthy_replicas == 0:
            # No healthy replicas, need to scale up
            return {"no_healthy": 10.0}
        
        # RPS per replica
        if policy.target_rps_per_replica > 0:
            rps_per_replica = metrics.rps / metrics.healthy_replicas
            factors["rps"] = rps_per_replica / policy.target_rps_per_replica
        
        # Latency
        if policy.max_p95_latency_ms > 0:
            factors["latency"] = metrics.p95_latency_ms / policy.max_p95_latency_ms
        
        # Connections per replica
        if policy.max_conn_per_replica > 0:
            conn_per_replica = metrics.active_connections / metrics.healthy_replicas
            factors["connections"] = conn_per_replica / policy.max_conn_per_replica
        
        # CPU
        if policy.max_cpu_percent > 0:
            factors["cpu"] = metrics.cpu_percent / policy.max_cpu_percent
        
        # Memory
        if policy.max_memory_percent > 0:
            factors["memory"] = metrics.memory_percent / policy.max_memory_percent
        
        return factors
    
    def _make_scaling_decision(
        self,
        app_name: str,
        current_replicas: int,
        scale_factors: Dict[str, float],
        policy: ScalingPolicy,
        metrics: ScalingMetrics
    ) -> ScalingDecision:
        """Make the final scaling decision based on scale factors."""
        
        triggered_by = []
        max_factor = 0.0
        min_factor = 1.0
        
        for metric_name, factor in scale_factors.items():
            max_factor = max(max_factor, factor)
            min_factor = min(min_factor, factor)
            
            # Check if this metric triggers scaling
            scale_out_threshold = policy.scale_out_threshold_pct / 100.0
            scale_in_threshold = policy.scale_in_threshold_pct / 100.0
            
            if factor > scale_out_threshold:
                triggered_by.append(f"{metric_name}={factor:.2f}")
        
        # Decide on scaling action
        target_replicas = current_replicas
        should_scale = False
        reason = "Metrics within thresholds"
        
        # Scale out if any metric is above threshold
        if max_factor > (policy.scale_out_threshold_pct / 100.0) and current_replicas < policy.max_replicas:
            # Calculate desired replicas based on the worst metric
            desired_replicas = int(current_replicas * max_factor) + 1
            target_replicas = min(desired_replicas, policy.max_replicas)
            should_scale = True
            reason = f"Scale out: max factor {max_factor:.2f}"
        
        # Scale in if all metrics are below threshold
        elif max_factor < (policy.scale_in_threshold_pct / 100.0) and current_replicas > policy.min_replicas:
            # Conservative scale-in: only reduce by 1 replica at a time
            target_replicas = max(current_replicas - 1, policy.min_replicas)
            should_scale = True
            reason = f"Scale in: max factor {max_factor:.2f}"
        
        # Special case: no healthy replicas
        if "no_healthy" in scale_factors:
            target_replicas = min(current_replicas + 1, policy.max_replicas)
            should_scale = True
            reason = "No healthy replicas available"
            triggered_by = ["no_healthy"]
        
        decision = ScalingDecision(
            should_scale=should_scale,
            target_replicas=target_replicas,
            current_replicas=current_replicas,
            reason=reason,
            triggered_by=triggered_by,
            metrics=metrics
        )
        
        if should_scale:
            logger.info(f"Scaling decision for app: {reason}, {current_replicas} -> {target_replicas}")
        else:
            logger.debug(f"[autoscale] No scale for {app_name}: reason={reason}, factors={scale_factors}")
        
        return decision
    
    def record_scaling_action(self, app_name: str, new_replicas: int):
        """Record that a scaling action was taken."""
        self.last_scale_time[app_name] = time.time()
        logger.info(f"Recorded scaling action for {app_name}: {new_replicas} replicas")
    
    def get_scaling_history(self, app_name: str, limit: int = 10) -> List[ScalingDecision]:
        """Get recent scaling decisions for an application."""
        decisions = list(self.scale_decisions[app_name])
        return decisions[-limit:] if decisions else []
    
    def get_metrics_summary(self, app_name: str) -> Dict:
        """Get a summary of recent metrics for an application."""
        policy = self.policies.get(app_name)
        if not policy:
            return {}
        
        recent_metrics = self._get_recent_metrics(app_name, policy.window_seconds)
        if not recent_metrics:
            return {}
        
        scale_factors = self._calculate_scale_factors(recent_metrics, policy)
        
        return {
            "metrics": {
                "rps": recent_metrics.rps,
                "p95_latency_ms": recent_metrics.p95_latency_ms,
                "active_connections": recent_metrics.active_connections,
                "cpu_percent": recent_metrics.cpu_percent,
                "memory_percent": recent_metrics.memory_percent,
                "healthy_replicas": recent_metrics.healthy_replicas
            },
            "scale_factors": scale_factors,
            "policy": {
                "min_replicas": policy.min_replicas,
                "max_replicas": policy.max_replicas,
                "target_rps_per_replica": policy.target_rps_per_replica,
                "max_p95_latency_ms": policy.max_p95_latency_ms,
                "scale_out_threshold_pct": policy.scale_out_threshold_pct,
                "scale_in_threshold_pct": policy.scale_in_threshold_pct
            }
        }

File: controller_go/scaler.go
package controller

import (
	"errors"
	"log"
	"math"
	"sort"
	"sync"
	"time"
)

const (
	MetricsRetentionMultiplier = 2
	MinScaleInStablePeriods    = 3
	EmergencyScaleFactor       = 10.0
)

type ScalingPolicy struct {
	MinReplicas          int
	MaxReplicas          int
	TargetRPSPerReplica  int
	MaxP95LatencyMs      int
	MaxConnPerReplica    int
	ScaleOutThresholdPct int
	ScaleInThresholdPct  int
	WindowSeconds        int
	CooldownSeconds      int
	MaxCPUPercent        float64
	MaxMemoryPercent     float64
}

func (p *ScalingPolicy) Validate() error {
	if p.MinReplicas < 1 {
		return errors.New("min_replicas must be >= 1")
	}
	if p.MaxReplicas < p.MinReplicas {
		return errors.New("max_replicas must be >= min_replicas")
	}
	if p.ScaleInThresholdPct >= p.ScaleOutThresholdPct {
		return errors.New("scale_in_threshold must be < scale_out_threshold")
	}
	if p.WindowSeconds < 1 {
		return errors.New("window_seconds must be >= 1")
	}
	if p.CooldownSeconds < 0 {
		return errors.New("cooldown_seconds must be >= 0")
	}
	return nil
}

type MetricPoint struct {
	Timestamp float64
	Value     float64
}

type ScalingMetrics struct {
	RPS               float64
	P95LatencyMs      float64
	ActiveConnections int
	CPUPercent        float64
	MemoryPercent     float64
	HealthyReplicas   int
	TotalReplicas     int
}

type ScalingDecision struct {
	ShouldScale     bool
	TargetReplicas  int
	CurrentReplicas int
	Reason          string
	TriggeredBy     []string
	Metrics         ScalingMetrics
}

type AutoScaler struct {
	mu                   sync.RWMutex
	policies             map[string]*ScalingPolicy
	metricsHistory       map[string]map[string][]MetricPoint
	lastScaleTime        map[string]float64
	scaleDecisions       map[string][]ScalingDecision
	lastScaleFactors     map[string]map[string]float64
	scaleInStablePeriods map[string]int
}

func NewAutoScaler() *AutoScaler {
	return &AutoScaler{
		policies:             make(map[string]*ScalingPolicy),
		metricsHistory:       make(map[string]map[string][]MetricPoint),
		lastScaleTime:        make(map[string]float64),
		scaleDecisions:       make(map[string][]ScalingDecision),
		lastScaleFactors:     make(map[string]map[string]float64),
		scaleInStablePeriods: make(map[string]int),
	}
}

func (a *AutoScaler) SetPolicy(app string, policy ScalingPolicy) error {
	if err := policy.Validate(); err != nil {
		return err
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	a.policies[app] = &policy
	log.Printf("[AutoScaler] Set policy for %s: min=%d max=%d out=%d%% in=%d%%",
		app, policy.MinReplicas, policy.MaxReplicas,
		policy.ScaleOutThresholdPct, policy.ScaleInThresholdPct)
	return nil
}

func (a *AutoScaler) AddMetrics(app string, metrics ScalingMetrics) {
	a.mu.Lock()
	defer a.mu.Unlock()

	ts := float64(time.Now().Unix())
	if _, ok := a.metricsHistory[app]; !ok {
		a.metricsHistory[app] = map[string][]MetricPoint{
			"rps":         {},
			"latency":     {},
			"connections": {},
			"cpu":         {},
			"memory":      {},
			"healthy":     {},
			"total":       {},
		}
	}

	h := a.metricsHistory[app]
	h["rps"] = append(h["rps"], MetricPoint{ts, metrics.RPS})
	h["latency"] = append(h["latency"], MetricPoint{ts, metrics.P95LatencyMs})
	h["connections"] = append(h["connections"], MetricPoint{ts, float64(metrics.ActiveConnections)})
	h["cpu"] = append(h["cpu"], MetricPoint{ts, metrics.CPUPercent})
	h["memory"] = append(h["memory"], MetricPoint{ts, metrics.MemoryPercent})
	h["healthy"] = append(h["healthy"], MetricPoint{ts, float64(metrics.HealthyReplicas)})
	h["total"] = append(h["total"], MetricPoint{ts, float64(metrics.TotalReplicas)})

	a.cleanOldMetrics(app, ts)
}

func (a *AutoScaler) cleanOldMetrics(app string, now float64) {
	policy, ok := a.policies[app]
	if !ok {
		return
	}
	cutoff := now - float64(policy.WindowSeconds*MetricsRetentionMultiplier)
	for metric, pts := range a.metricsHistory[app] {
		var kept []MetricPoint
		for _, p := range pts {
			if p.Timestamp >= cutoff {
				kept = append(kept, p)
			}
		}
		a.metricsHistory[app][metric] = kept
	}
}

func avg(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range vals {
		sum += v
	}
	return sum / float64(len(vals))
}

func percentile(vals []float64, p float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	sort.Float64s(vals)
	k := int(float64(len(vals)-1) * p)
	return vals[k]
}

func (a *AutoScaler) getRecentMetrics(app string, window int) *ScalingMetrics {
	now := float64(time.Now().Unix())
	cutoff := now - float64(window)
	history, ok := a.metricsHistory[app]
	if !ok {
		return nil
	}

	filter := func(points []MetricPoint) []float64 {
		var vals []float64
		for _, p := range points {
			if p.Timestamp >= cutoff {
				vals = append(vals, p.Value)
			}
		}
		return vals
	}

	rpsVals := filter(history["rps"])
	latencyVals := filter(history["latency"])
	connVals := filter(history["connections"])
	cpuVals := filter(history["cpu"])
	memVals := filter(history["memory"])
	healthyVals := filter(history["healthy"])
	totalVals := filter(history["total"])

	if len(rpsVals) == 0 && len(latencyVals) == 0 {
		return nil
	}

	p95 := percentile(latencyVals, 0.95)
	return &ScalingMetrics{
		RPS:               avg(rpsVals),
		P95LatencyMs:      p95,
		ActiveConnections: int(avg(connVals)),
		CPUPercent:        avg(cpuVals),
		MemoryPercent:     avg(memVals),
		HealthyReplicas:   int(math.Max(1, avg(healthyVals))),
		TotalReplicas:     int(math.Max(1, avg(totalVals))),
	}
}

func (a *AutoScaler) Evaluate(app string, current int, mode string) ScalingDecision {
	a.mu.Lock()
	defer a.mu.Unlock()

	if mode == "manual" {
		return ScalingDecision{ShouldScale: false, TargetReplicas: current, CurrentReplicas: current, Reason: "Manual mode"}
	}

	policy, ok := a.policies[app]
	if !ok {
		return ScalingDecision{ShouldScale: false, TargetReplicas: current, CurrentReplicas: current, Reason: "No policy"}
	}

	if current < policy.MinReplicas {
		a.scaleInStablePeriods[app] = 0
		return ScalingDecision{
			ShouldScale:     true,
			TargetReplicas:  policy.MinReplicas,
			CurrentReplicas: current,
			Reason:          "Below min replicas",
			TriggeredBy:     []string{"min_replicas_enforcement"},
		}
	}

	last := a.lastScaleTime[app]
	if time.Since(time.Unix(int64(last), 0)) < time.Duration(policy.CooldownSeconds)*time.Second {
		return ScalingDecision{ShouldScale: false, TargetReplicas: current, CurrentReplicas: current, Reason: "In cooldown"}
	}

	metrics := a.getRecentMetrics(app, policy.WindowSeconds)
	if metrics == nil {
		return ScalingDecision{ShouldScale: false, TargetReplicas: current, CurrentReplicas: current, Reason: "No metrics"}
	}

	factors := a.calculateScaleFactors(*metrics, policy)
	a.lastScaleFactors[app] = factors

	decision := a.makeScalingDecision(app, current, factors, policy, *metrics)
	a.scaleDecisions[app] = append(a.scaleDecisions[app], decision)
	return decision
}

func (a *AutoScaler) calculateScaleFactors(metrics ScalingMetrics, policy *ScalingPolicy) map[string]float64 {
	factors := map[string]float64{}

	if metrics.HealthyReplicas == 0 {
		factors["no_healthy"] = EmergencyScaleFactor
		return factors
	}

	if policy.TargetRPSPerReplica > 0 {
		factors["rps"] = (metrics.RPS / float64(metrics.HealthyReplicas)) / float64(policy.TargetRPSPerReplica)
	}
	if policy.MaxP95LatencyMs > 0 && metrics.P95LatencyMs > 0 {
		factors["latency"] = metrics.P95LatencyMs / float64(policy.MaxP95LatencyMs)
	}
	if policy.MaxConnPerReplica > 0 {
		factors["connections"] = (float64(metrics.ActiveConnections) / float64(metrics.HealthyReplicas)) / float64(policy.MaxConnPerReplica)
	}
	if policy.MaxCPUPercent > 0 {
		factors["cpu"] = metrics.CPUPercent / policy.MaxCPUPercent
	}
	if policy.MaxMemoryPercent > 0 {
		factors["memory"] = metrics.MemoryPercent / policy.MaxMemoryPercent
	}
	return factors
}

func (a *AutoScaler) makeScalingDecision(app string, current int, factors map[string]float64, policy *ScalingPolicy, metrics ScalingMetrics) ScalingDecision {
	triggered := []string{}
	maxFactor := 0.0
	scaleOutThreshold := float64(policy.ScaleOutThresholdPct) / 100.0

	for k, f := range factors {
		if f > maxFactor {
			maxFactor = f
		}
		if f > scaleOutThreshold {
			triggered = append(triggered, k)
		}
	}

	target := current
	shouldScale := false
	reason := "Metrics within thresholds"

	if _, ok := factors["no_healthy"]; ok {
		target = int(math.Min(float64(current+1), float64(policy.MaxReplicas)))
		return ScalingDecision{ShouldScale: true, TargetReplicas: target, CurrentReplicas: current, Reason: "No healthy replicas", TriggeredBy: []string{"no_healthy"}, Metrics: metrics}
	}

	// Scale out
	if maxFactor > scaleOutThreshold && current < policy.MaxReplicas {
		desired := int(math.Ceil(float64(current) * maxFactor))
		if desired <= current {
			desired = current + 1
		}
		target = int(math.Min(float64(desired), float64(policy.MaxReplicas)))
		shouldScale = target > current
		reason = "Scale out triggered"
		a.scaleInStablePeriods[app] = 0
	}

	// Scale in
	if maxFactor < float64(policy.ScaleInThresholdPct)/100.0 && current > policy.MinReplicas {
		a.scaleInStablePeriods[app]++
		if a.scaleInStablePeriods[app] >= MinScaleInStablePeriods {
			target = int(math.Max(float64(current-1), float64(policy.MinReplicas)))
			shouldScale = target < current
			reason = "Scale in triggered"
			a.scaleInStablePeriods[app] = 0
		}
	}

	return ScalingDecision{ShouldScale: shouldScale, TargetReplicas: target, CurrentReplicas: current, Reason: reason, TriggeredBy: triggered, Metrics: metrics}
}

// EvaluateScaling is an alias for Evaluate for compatibility with API
func (a *AutoScaler) EvaluateScaling(app string, current int, mode string) ScalingDecision {
	return a.Evaluate(app, current, mode)
}

// RecordScalingAction records that a scaling action occurred
func (a *AutoScaler) RecordScalingAction(app string, replicas int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.lastScaleTime[app] = float64(time.Now().Unix())
}

// GetMetricsSummary returns a summary of metrics for an app
func (a *AutoScaler) GetMetricsSummary(app string) map[string]interface{} {
	a.mu.RLock()
	defer a.mu.RUnlock()

	policy, ok := a.policies[app]
	if !ok {
		return map[string]interface{}{
			"error": "No policy set for app",
		}
	}

	metricsHistory, ok := a.metricsHistory[app]
	if !ok || len(metricsHistory) == 0 {
		return map[string]interface{}{
			"policy":         policy,
			"recent_metrics": nil,
			"window_seconds": policy.WindowSeconds,
		}
	}

	recentMetrics := a.getRecentMetrics(app, policy.WindowSeconds)

	return map[string]interface{}{
		"policy":          policy,
		"recent_metrics":  recentMetrics,
		"window_seconds":  policy.WindowSeconds,
		"last_scale_time": a.lastScaleTime[app],
	}
}

// GetLastScaleFactors returns the last calculated scale factors for an app
func (a *AutoScaler) GetLastScaleFactors(app string) map[string]float64 {
	a.mu.RLock()
	defer a.mu.RUnlock()

	if factors, ok := a.lastScaleFactors[app]; ok {
		return factors
	}
	return map[string]float64{}
}

// GetPolicy returns the scaling policy for an app
func (a *AutoScaler) GetPolicy(app string) *ScalingPolicy {
	a.mu.RLock()
	defer a.mu.RUnlock()

	return a.policies[app] // policies map already stores pointers, so return directly
}
